e2e/testcases/otel_collector_test.go

```diff
@@ -0,0 +1,150 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"kpt.dev/configsync/e2e"
+	"kpt.dev/configsync/e2e/nomostest"
+	"kpt.dev/configsync/e2e/nomostest/ntopts"
+	"kpt.dev/configsync/e2e/nomostest/taskgroup"
+	nomostesting "kpt.dev/configsync/e2e/nomostest/testing"
+	"kpt.dev/configsync/pkg/kinds"
+	ocmetrics "kpt.dev/configsync/pkg/metrics"
+)
+
+const (
+	DefaultMonitorKSA = "default"
+	MonitorGSA        = "e2e-test-metric-writer"
+)
+
+func TestOtelCollectorDeployment(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation1, ntopts.RequireGKE(t))
+	nt.T.Cleanup(func() {
+		nt.MustKubectl("delete", "-f", "../testdata/otel-collector/otel-cm-duplicate-timeseries.yaml", "--ignore-not-found")
+		if t.Failed() {
+			nt.PodLogs("config-management-monitoring", ocmetrics.OtelCollectorName, "", false)
+		}
+	})
+
+	// If Workload Identity enabled on cluster, setup KSA to GSA annotation
+	if workloadPool, err := getWorkloadPool(nt); err != nil {
+		nt.T.Fatal(err)
+	} else if workloadPool != "" {
+		nt.T.Log("Workload identity enabled, adding KSA annotation to use %s service account", MonitorGSA)
+		nt.MustKubectl("annotate", "serviceaccount", "--namespace", ocmetrics.MonitoringNamespace, DefaultMonitorKSA,
+			fmt.Sprintf("iam.gke.io/gcp-service-account=%s@%s.iam.gserviceaccount.com", MonitorGSA, *e2e.GCPProject))
+	}
+
+	nt.T.Log("Restart otel-collector pod to refresh the ConfigMap, log and IAM")
+	nomostest.DeletePodByLabel(nt, "app", "opentelemetry", false)
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := validateOtelCollectorStatusCurrent(nt, false); err != nil {
+		nt.T.Fatal()
+	}
+
+	nt.T.Log("Wait 1 minute for metric exporter to settle")
+	time.Sleep(1 * time.Minute)
+	nt.T.Log("Checking the otel-collector log contains no failure...")
+	err := CheckDeploymentLogHasNoFailure(nt, ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("Apply custom otel-collector ConfigMap that could cause duplicate time series error")
+	nt.MustKubectl("apply", "-f", "../testdata/otel-collector/otel-cm-duplicate-timeseries.yaml")
+	nt.T.Log("Restart otel-collector pod to refresh the ConfigMap and log")
+	nomostest.DeletePodByLabel(nt, "app", "opentelemetry", false)
+	if err := nt.WatchForAllSyncs(); err != nil {
+		nt.T.Fatal(err)
+	}
+	if err := validateOtelCollectorStatusCurrent(nt, true); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.T.Log("wait 1 minute for metric exporter to settle")
+	time.Sleep(1 * time.Minute)
+	nt.T.Log("Checking the otel-collector log contains failure...")
+	if err := CheckDeploymentLogHasFailure(nt, ocmetrics.OtelCollectorName, ocmetrics.MonitoringNamespace); err != nil {
+		nt.T.Fatal(err)
+	}
+}
+
+func CheckDeploymentLogHasFailure(nt *nomostest.NT, deployment, namespace string) error {
```

Comment:
nit: this function is named pretty generically, so I'd suggest surfacing the log string (used by `strings.Contains`) as a parameter

Reply:
Surfaced failure string to args.